bot/src/keeper/surplus.ts

```diff
@@ -0,0 +1,109 @@
+import { SurplusAuctionActive } from 'auctions-core/src/types';
+import { bidToSurplusAuction, enrichSurplusAuction } from 'auctions-core/src/surplus';
+import getSigner from 'auctions-core/src/signer';
+import { fetchAllowanceAmountMKR, setAllowanceAmountMKR } from 'auctions-core/src/authorizations';
+import { KEEPER_MINIMUM_NET_PROFIT_DAI } from '../variables';
+import { isSetupCompleted } from './setup';
+import BigNumber from '~/../core/src/bignumber';
+
+const currentlyExecutedAuctions = new Set();
+
+const checkAndParticipateIfPossible = async function (network: string, auction: SurplusAuctionActive) {
+    // check if setupKeeper hasn't run
+    if (!isSetupCompleted) {
+        return;
+    }
+
+    const signer = await getSigner(network);
+
+    // enrich the auction with more numbers
+    const auctionTransaction = await enrichSurplusAuction(network, auction);
+
+    // check if auction became inactive or finished
+    if (auctionTransaction.state === 'ready-for-collection') {
+        console.info(`keeper: surplus auction "${auction.id}" has already finished`);
+        return;
+    }
+    if (auctionTransaction.state === 'requires-restart') {
+        console.info(`keeper: surplus auction "${auction.id}" is inactive`);
+        return;
+    }
+
+    // check auction's profit
+    const profit = new BigNumber(1)
+        .div(auctionTransaction.marketUnitPrice.minus(auctionTransaction.unitPrice))
+        .times(auctionTransaction.bidAmountMKR);
+    if (profit.isLessThan(0)) {
+        console.info(
+            `keeper: surplus auction "${auction.id}" is not yet profitable (current profit: ${profit.toFixed(0)} DAI)`
+        );
+        return;
+    }
+
+    // check auction's clear profit – profit without transaction fees
+    const clearProfit = profit.minus(auctionTransaction.combinedBidFeesDai);
+    if (clearProfit.isLessThan(new BigNumber(KEEPER_MINIMUM_NET_PROFIT_DAI))) {
+        console.info(
+            `keeper: surplus auction "${auction.id}" clear profit is smaller than min profit (${clearProfit.toFixed(
+                0
+            )} < ${KEEPER_MINIMUM_NET_PROFIT_DAI})`
+        );
+        return;
+    } else {
+        console.info(
+            `keeper: surplus auction "${auction.id}" clear profit is ${clearProfit.toFixed(
+                0
+            )} DAI after transaction fees, checking wallet MKR allowance`
+        );
+    }
+
+    const walletAddress = await signer.getAddress();
+
+    // check the wallet's MKR allowance
```

Comment:
Don't we want to check how much MKR in VAT the wallet holds before we set the allowance?

If not enough is present in the VAT we can also deposit any MKR into the VAT, the wallet holds.